operatorcourier/validate.py

```diff
@@ -167,20 +172,55 @@ def _csv_spec_validation(self, spec, bundleData):
                     pass
 
             if "owned" not in customresourcedefinitions:
-                self._log_error("spec.customresourcedefinitions.owned"
+                self._log_error("spec.customresourcedefinitions.owned "
                                 "not defined for csv")
-                valid = False
-            else:
-                for crd in customresourcedefinitions["owned"]:
-                    if "name" not in crd:
-                        self._log_error("name not defined for item in "
-                                        "spec.customresourcedefinitions.")
-                        valid = False
-                    else:
-                        if crd["name"] not in crdList:
-                            self._log_error("custom resource definition referenced "
-                                            "in csv not defined in root list of crds")
+                return False
+
+            for csvOwnedCrd in customresourcedefinitions["owned"]:
+                if "name" not in csvOwnedCrd:
+                    self._log_error("name not defined for item in "
+                                    "spec.customresourcedefinitions.")
+                    valid = False
+                elif csvOwnedCrd["name"] not in crdList:
+                    self._log_error("custom resource definition %s referenced in csv "
+                                    "not defined in root list of crds",
+                                    csvOwnedCrd["name"])
+                    valid = False
 
+                if "kind" not in csvOwnedCrd:
+                    self._log_error("kind not defined for item in "
+                                    "spec.customresourcedefinitions.")
+                    valid = False
+                if "version" not in csvOwnedCrd:
+                    self._log_error("version not defined for item in "
+                                    "spec.customresourcedefinitions.")
+                    valid = False
+
+                for crd in bundleData[self.crdKey]:
+                    if 'name' not in csvOwnedCrd or not \
+                        self._crd_fields_validation(['metadata', 'name'], crd, '') \
```

Comment:
If a CRD in the bundle doesn't have a name, should that also log as an error?

Reply:
Hi @SamiSousa, thanks for the review. Yes, but I believe this should be covered in [previous checks](https://github.com/operator-framework/operator-courier/blob/25af109b1a25a74ab805e489be9981ee413ada9a/operatorcourier/validate.py#L180-L183).

Reply:
That check is for the CRD in the CSV, rather than the CRD from the bundle directly.
https://github.com/operator-framework/operator-courier/blob/25af109b1a25a74ab805e489be9981ee413ada9a/operatorcourier/validate.py#L166
On this line the `crdList` is created. Should this be used in this for loop?

Reply:
Yeah, for each type we should have a list of fields we check. The fact that they happen to be checked in other places by code that compares those fields to other fields shouldn't be relevant. I think part of the problem is that adding this required fields list `crd_required_nested_fields` makes the code hard to read.

Perhaps instead we should either do that everywhere for every type, or not create that abstraction at all. Either way, it actually seems outside the scope of this pr. Can we revert it?

Reply:
@kevinrizza Without this abstraction we will need to introduce even more nested `if`s just to check the presence of nested fields. These are already checked in CRD checks but we need to do it again to make sure the field in the yaml we compare against csvOwnedCrd actually exists.

Maybe after this PR I can apply this abstraction for checks in all types? And this hopefully can reduce the nested levels like [this](https://github.com/operator-framework/operator-courier/blob/25af109b1a25a74ab805e489be9981ee413ada9a/operatorcourier/validate.py#L276-L307) that can be found in multiple places in this file.

Reply:
Your intuition about nested `if`s is good in the general case, but I think in this case it makes the code more clear. If we abstract away certain steps but only in certain cases, it makes the code hard to reason about.

Additionally, I think ultimately implementing a change like the one described here is probably the better long term approach: https://github.com/operator-framework/operator-courier/issues/39

Reply:
@kevinrizza That seems really cool. For this particular PR, what are the changes you want me to make? Should I remove this abstraction and replace it with nested `if`s instead?

Reply:
@jeremylinlin yes, and please ensure that any of the checks were doing before are no longer removed.

Reply:
/cc @kevinrizza @SamiSousa Updated.